File: cozip/python/tests/test_smoke.py
"""Conformance tests for the cozip Python writer"""

import io
import struct
import zipfile
from pathlib import Path

import pyarrow as pa
import pyarrow.parquet as pq
import pytest

import cozip


SMALL_CONTENT = b"hello cozip\n" * 8
MEDIUM_CONTENT = bytes(range(256)) * 160

INDEX_OFFSET = 51
HASH_WINDOW = 32768
ZIP_LFH_SIG = b"PK\x03\x04"
INDEX_NAME = b"__cozip__"
HASH_BLOCK_HEADER = b"\x0c\xca\x08\x00"
INDEX_MAGIC = b"CZIP"
INDEX_VERSION = b"\x01\x00"
PROFILE_FLAT = 1


def fnv1a_64(data: bytes) -> int:
    h = 0xCBF29CE484222325
    prime = 0x100000001B3
    for b in data:
        h ^= b
        h = (h * prime) & 0xFFFFFFFFFFFFFFFF
    return h


def hash_input(archive: bytes, index_size: int) -> bytes:
    archive_size = len(archive)
    suffix_start = archive_size - HASH_WINDOW
    index_end = INDEX_OFFSET + index_size
    if index_end <= suffix_start:
        return archive[INDEX_OFFSET:index_end] + archive[suffix_start:archive_size]
    return archive[INDEX_OFFSET:archive_size]


def parse_index(payload: bytes) -> dict[str, tuple[int, int]]:
    assert payload[:4] == INDEX_MAGIC, f"bad magic: {payload[:4]!r}"
    n = struct.unpack_from("<I", payload, 7)[0]
    cur = 11
    name_lens = struct.unpack_from(f"<{n}H", payload, cur)
    cur += 2 * n
    names = []
    for nl in name_lens:
        names.append(payload[cur:cur + nl].decode("utf-8"))
        cur += nl
    offsets = struct.unpack_from(f"<{n}Q", payload, cur)
    cur += 8 * n
    sizes = struct.unpack_from(f"<{n}Q", payload, cur)
    return {nm: (off, sz) for nm, off, sz in zip(names, offsets, sizes)}


def index_size_from_lfh(archive: bytes) -> int:
    return struct.unpack("<I", archive[18:22])[0]


def extract_metadata_payload(archive: bytes) -> bytes:
    payload = archive[INDEX_OFFSET:INDEX_OFFSET + index_size_from_lfh(archive)]
    offset, size = parse_index(payload)["__metadata__"]
    return archive[offset:offset + size]


def assert_valid_cozip(data: bytes) -> None:
    assert len(data) >= HASH_WINDOW + INDEX_OFFSET
    assert data[:4] == ZIP_LFH_SIG
    assert data[30:39] == INDEX_NAME
    assert data[51:55] == INDEX_MAGIC
    expected = fnv1a_64(hash_input(data, index_size_from_lfh(data)))
    stored = struct.unpack("<Q", data[43:51])[0]
    assert stored == expected


@pytest.fixture
def fixtures(tmp_path: Path) -> dict[str, Path]:
    small = tmp_path / "small.txt"
    small.write_bytes(SMALL_CONTENT)
    medium = tmp_path / "medium.bin"
    medium.write_bytes(MEDIUM_CONTENT)
    return {"small": small, "medium": medium}


@pytest.fixture
def input_table(fixtures: dict[str, Path]) -> pa.Table:
    return pa.table({
        "name": ["a.txt", "b.bin"],
        "path": [str(fixtures["small"]), str(fixtures["medium"])],
        "category": ["text", "binary"],
    })


@pytest.fixture
def paths_arg(input_table: pa.Table) -> list[tuple[str, str]]:
    return list(zip(
        input_table.column("name").to_pylist(),
        input_table.column("path").to_pylist(),
    ))


@pytest.fixture
def archive(tmp_path: Path, input_table: pa.Table) -> Path:
    out = tmp_path / "out.zip"
    cozip.create(out, input_table)
    return out


class TestSpecInvariants:
    def test_archive_size_minimum(self, archive: Path) -> None:
        assert archive.stat().st_size >= HASH_WINDOW + INDEX_OFFSET

    def test_lfh_signature(self, archive: Path) -> None:
        assert archive.read_bytes()[:4] == ZIP_LFH_SIG

    def test_index_entry_filename(self, archive: Path) -> None:
        assert archive.read_bytes()[30:39] == INDEX_NAME

    def test_hash_block_header(self, archive: Path) -> None:
        assert archive.read_bytes()[39:43] == HASH_BLOCK_HEADER

    def test_index_header(self, archive: Path) -> None:
        data = archive.read_bytes()
        assert data[51:55] == INDEX_MAGIC
        assert data[55:57] == INDEX_VERSION
        assert data[57] == PROFILE_FLAT

    def test_eocd_comment_is_empty(self, archive: Path) -> None:
        assert archive.read_bytes()[-2:] == b"\x00\x00"

    def test_integrity_hash(self, archive: Path) -> None:
        data = archive.read_bytes()
        expected = fnv1a_64(hash_input(data, index_size_from_lfh(data)))
        stored = struct.unpack("<Q", data[43:51])[0]
        assert stored == expected, (
            f"stored=0x{stored:016x} expected=0x{expected:016x}"
        )

    def test_index_lists_only_metadata(self, archive: Path) -> None:
        data = archive.read_bytes()
        entries = parse_index(data[INDEX_OFFSET:INDEX_OFFSET + index_size_from_lfh(data)])
        assert set(entries) == {"__metadata__"}
        offset, size = entries["__metadata__"]
        assert 0 < offset < len(data)
        assert 0 < size <= len(data) - offset


class TestZipCompatibility:
    def test_stdlib_zipfile_lists_all_entries(self, archive: Path) -> None:
        with zipfile.ZipFile(archive) as zf:
            names = set(zf.namelist())
        assert {"__cozip__", "__metadata__", "a.txt", "b.bin"} <= names

    def test_all_entries_use_store(self, archive: Path) -> None:
        with zipfile.ZipFile(archive) as zf:
            for info in zf.infolist():
                assert info.compress_type == zipfile.ZIP_STORED, info.filename

    def test_no_encryption_no_data_descriptor(self, archive: Path) -> None:
        with zipfile.ZipFile(archive) as zf:
            for info in zf.infolist():
                assert info.flag_bits & 0x01 == 0, f"{info.filename} encrypted"
                assert info.flag_bits & 0x08 == 0, f"{info.filename} has DD"

    def test_user_payloads_byte_exact(self, archive: Path) -> None:
        with zipfile.ZipFile(archive) as zf:
            assert zf.read("a.txt") == SMALL_CONTENT
            assert zf.read("b.bin") == MEDIUM_CONTENT


class TestStageMetadata:
    def test_drops_path_column(self, input_table: pa.Table) -> None:
        meta, _ = cozip.stage_metadata(input_table)
        assert "path" not in meta.column_names

    def test_adds_offset_and_size_as_uint64(self, input_table: pa.Table) -> None:
        meta, _ = cozip.stage_metadata(input_table)
        assert meta.schema.field("offset").type == pa.uint64()
        assert meta.schema.field("size").type == pa.uint64()

    def test_preserves_user_extras(self, input_table: pa.Table) -> None:
        meta, _ = cozip.stage_metadata(input_table)
        assert meta.column("category").to_pylist() == ["text", "binary"]

    def test_canonical_column_order(self, input_table: pa.Table) -> None:
        meta, _ = cozip.stage_metadata(input_table)
        assert meta.column_names[:3] == ["name", "offset", "size"]

    def test_sizes_match_source_lengths(self, input_table: pa.Table) -> None:
        meta, _ = cozip.stage_metadata(input_table)
        assert meta.column("size").to_pylist() == [len(SMALL_CONTENT), len(MEDIUM_CONTENT)]

    def test_offsets_strictly_increasing(self, input_table: pa.Table) -> None:
        meta, _ = cozip.stage_metadata(input_table)
        offsets = meta.column("offset").to_pylist()
        assert all(a < b for a, b in zip(offsets, offsets[1:]))

    def test_returns_paths_aligned_with_table(self, input_table: pa.Table) -> None:
        meta, paths = cozip.stage_metadata(input_table)
        expected = list(zip(
            input_table.column("name").to_pylist(),
            input_table.column("path").to_pylist(),
        ))
        assert paths == expected
        assert len(paths) == meta.num_rows

    def test_paths_drive_stage_create(
        self, tmp_path: Path, input_table: pa.Table
    ) -> None:
        meta, paths = cozip.stage_metadata(input_table)
        meta_pq = tmp_path / "meta.parquet"
        pq.write_table(meta, meta_pq)
        out = tmp_path / "out.zip"
        cozip.stage_create(out, paths, meta_pq)
        assert_valid_cozip(out.read_bytes())

    def test_rejects_offset_in_input(self, fixtures: dict[str, Path]) -> None:
        bad = pa.table({
            "name": ["a.txt"],
            "path": [str(fixtures["small"])],
            "offset": [0],
        })
        with pytest.raises(ValueError, match="reserved"):
            cozip.stage_metadata(bad)

    def test_rejects_duplicate_names(self, fixtures: dict[str, Path]) -> None:
        bad = pa.table({
            "name": ["x", "x"],
            "path": [str(fixtures["small"]), str(fixtures["medium"])],
        })
        with pytest.raises(ValueError, match="duplicate"):
            cozip.stage_metadata(bad)

    def test_rejects_reserved_archive_name(self, fixtures: dict[str, Path]) -> None:
        bad = pa.table({
            "name": ["__metadata__"],
            "path": [str(fixtures["small"])],
        })
        with pytest.raises(ValueError, match="reserved"):
            cozip.stage_metadata(bad)


class TestStageCreate:
    def _write_meta(self, tmp_path: Path, input_table: pa.Table) -> Path:
        meta, _ = cozip.stage_metadata(input_table)
        meta_pq = tmp_path / "meta.parquet"
        pq.write_table(meta, meta_pq)
        return meta_pq

    def test_packs_a_valid_archive(
        self, tmp_path: Path, input_table: pa.Table, paths_arg
    ) -> None:
        out = tmp_path / "out.zip"
        meta_pq = self._write_meta(tmp_path, input_table)
        cozip.stage_create(out, paths_arg, meta_pq)
        assert_valid_cozip(out.read_bytes())

    def test_metadata_parquet_embedded_verbatim(
        self, tmp_path: Path, input_table: pa.Table, paths_arg
    ) -> None:
        meta_pq = self._write_meta(tmp_path, input_table)
        original_bytes = meta_pq.read_bytes()

        out = tmp_path / "out.zip"
        cozip.stage_create(out, paths_arg, meta_pq)

        embedded = extract_metadata_payload(out.read_bytes())
        assert embedded == original_bytes

    def test_rejects_parquet_with_path_column(
        self, tmp_path: Path, input_table: pa.Table, paths_arg
    ) -> None:
        meta, _ = cozip.stage_metadata(input_table)
        tampered = meta.append_column("path", pa.array(["/tmp/a", "/tmp/b"]))
        meta_pq = tmp_path / "meta.parquet"
        pq.write_table(tampered, meta_pq)

        with pytest.raises(ValueError, match="path"):
            cozip.stage_create(tmp_path / "out.zip", paths_arg, meta_pq)

    def test_rejects_parquet_missing_required_columns(
        self, tmp_path: Path, paths_arg
    ) -> None:
        bad = pa.table({"name": ["a.txt", "b.bin"]})
        meta_pq = tmp_path / "meta.parquet"
        pq.write_table(bad, meta_pq)

        with pytest.raises(ValueError, match="required column"):
            cozip.stage_create(tmp_path / "out.zip", paths_arg, meta_pq)

    def test_rejects_offset_mismatch(
        self, tmp_path: Path, input_table: pa.Table, paths_arg
    ) -> None:
        meta, _ = cozip.stage_metadata(input_table)
        idx = meta.column_names.index("offset")
        bad = meta.set_column(idx, "offset", pa.array([0, 0], type=pa.uint64()))
        meta_pq = tmp_path / "meta.parquet"
        pq.write_table(bad, meta_pq)

        with pytest.raises(ValueError, match="offset mismatch"):
            cozip.stage_create(tmp_path / "out.zip", paths_arg, meta_pq)

    def test_rejects_row_count_mismatch(
        self, tmp_path: Path, input_table: pa.Table, paths_arg
    ) -> None:
        meta, _ = cozip.stage_metadata(input_table)
        truncated = meta.slice(0, 1)
        meta_pq = tmp_path / "meta.parquet"
        pq.write_table(truncated, meta_pq)

        with pytest.raises(ValueError, match="rows"):
            cozip.stage_create(tmp_path / "out.zip", paths_arg, meta_pq)


class TestSchemaMetadataRoundTrip:
    def test_arrow_schema_metadata_is_preserved(
        self, tmp_path: Path, input_table: pa.Table, paths_arg
    ) -> None:
        geo_value = (
            b'{"version":"1.0.0","primary_column":"geometry",'
            b'"columns":{"geometry":{"encoding":"WKB","crs":null}}}'
        )
        meta, _ = cozip.stage_metadata(input_table)
        tagged = meta.replace_schema_metadata(
            {b"geo": geo_value, b"asterisk:tag": b"cozip-conformance"}
        )
        meta_pq = tmp_path / "meta.parquet"
        pq.write_table(tagged, meta_pq)

        out = tmp_path / "out.zip"
        cozip.stage_create(out, paths_arg, meta_pq)

        embedded = extract_metadata_payload(out.read_bytes())
        recovered = pq.read_table(io.BytesIO(embedded))

        assert recovered.schema.metadata is not None
        assert recovered.schema.metadata.get(b"geo") == geo_value
        assert recovered.schema.metadata.get(b"asterisk:tag") == b"cozip-conformance"


class TestGeopandasRoundTrip:
    def test_geopandas_reads_embedded_geoparquet(
        self, tmp_path: Path, input_table: pa.Table, paths_arg
    ) -> None:
        gpd = pytest.importorskip("geopandas")
        from shapely.geometry import Point

        meta, _ = cozip.stage_metadata(input_table)
        df = meta.to_pandas()
        gdf = gpd.GeoDataFrame(
            df,
            geometry=[Point(-77.04, -12.05), Point(2.35, 48.86)],
            crs="EPSG:4326",
        )
        meta_pq = tmp_path / "meta.parquet"
        gdf.to_parquet(meta_pq)

        out = tmp_path / "out.zip"
        cozip.stage_create(out, paths_arg, meta_pq)

        extracted = tmp_path / "extracted.parquet"
        extracted.write_bytes(extract_metadata_payload(out.read_bytes()))
        recovered = gpd.read_parquet(extracted)

        assert isinstance(recovered, gpd.GeoDataFrame)
        assert recovered.crs.to_string() == "EPSG:4326"
        assert recovered["name"].tolist() == ["a.txt", "b.bin"]
        assert recovered["offset"].tolist() == df["offset"].tolist()
        assert recovered["size"].tolist() == df["size"].tolist()

        first, second = recovered.geometry.iloc[0], recovered.geometry.iloc[1]
        assert (first.x, first.y) == (-77.04, -12.05)
        assert (second.x, second.y) == (2.35, 48.86)

File: cozip/r/tests/testthat/test-smoke.R
# Mirror of @pytest.fixture archive.
local_archive <- function(envir = parent.frame()) {
  tmp <- withr::local_tempdir(.local_envir = envir)
  fix <- make_fixtures(tmp)
  tbl <- make_input_table(fix)
  arc <- make_archive(tmp, tbl)
  list(
    tmp_path      = tmp,
    fixtures      = fix,
    input_table   = tbl,
    archive       = arc,
    archive_bytes = readBin(arc, raw(), n = file.size(arc))
  )
}

# R's utils::unzip doesn't expose GP flags or compression method,
# so walk the Central Directory by hand.
parse_cd <- function(bytes) {
  n <- length(bytes)

  # EOCD: last 22 bytes, no archive comment.
  eocd <- n - 21L
  cd_size   <- read_u32_le(bytes[(eocd + 12L):(eocd + 15L)])
  cd_offset <- read_u32_le(bytes[(eocd + 16L):(eocd + 19L)])

  cur <- cd_offset + 1L
  end <- cd_offset + cd_size
  entries <- list()
  CD_SIG <- as.raw(c(0x50, 0x4B, 0x01, 0x02))

  while (cur <= end) {
    stopifnot(identical(bytes[cur:(cur + 3L)], CD_SIG))
    flag    <- read_u16_le(bytes[(cur +  8L):(cur +  9L)])
    method  <- read_u16_le(bytes[(cur + 10L):(cur + 11L)])
    fn_len  <- read_u16_le(bytes[(cur + 28L):(cur + 29L)])
    ex_len  <- read_u16_le(bytes[(cur + 30L):(cur + 31L)])
    cm_len  <- read_u16_le(bytes[(cur + 32L):(cur + 33L)])
    fn      <- rawToChar(bytes[(cur + 46L):(cur + 46L + fn_len - 1L)])
    entries[[length(entries) + 1L]] <- list(
      filename = fn, flag_bits = flag, compress_method = method
    )
    cur <- cur + 46L + fn_len + ex_len + cm_len
  }
  entries
}


describe("Spec invariants — bytes of create() output", {

  it("archive size meets minimum", {
    ctx <- local_archive()
    expect_gte(file.size(ctx$archive), HASH_WINDOW + INDEX_OFFSET)
  })

  it("LFH signature", {
    ctx <- local_archive()
    expect_identical(ctx$archive_bytes[1:4], ZIP_LFH_SIG)
  })

  it("index entry filename", {
    ctx <- local_archive()
    expect_identical(ctx$archive_bytes[31:39], INDEX_NAME)
  })

  it("hash block header", {
    ctx <- local_archive()
    expect_identical(ctx$archive_bytes[40:43], HASH_BLOCK_HEADER)
  })

  it("index header (magic + version + profile)", {
    ctx <- local_archive()
    b <- ctx$archive_bytes
    expect_identical(b[52:55], INDEX_MAGIC)
    expect_identical(b[56:57], INDEX_VERSION)
    expect_equal(as.integer(b[58]), PROFILE_FLAT)
  })

  it("EOCD comment is empty", {
    ctx <- local_archive()
    n <- length(ctx$archive_bytes)
    expect_identical(ctx$archive_bytes[(n - 1L):n], as.raw(c(0x00, 0x00)))
  })

  it("integrity hash matches FNV-1a 64", {
    ctx <- local_archive()
    b  <- ctx$archive_bytes
    sz <- index_size_from_lfh(b)
    expected <- fnv1a_64(hash_input(b, sz))
    sb <- b[44:51]
    stored <- c(hi = read_u32_le(sb[5:8]), lo = read_u32_le(sb[1:4]))
    expect_identical(stored, expected)
  })

  it("index lists exactly __metadata__", {
    ctx <- local_archive()
    b  <- ctx$archive_bytes
    sz <- index_size_from_lfh(b)
    entries <- parse_index(b[(INDEX_OFFSET + 1L):(INDEX_OFFSET + sz)])
    expect_setequal(names(entries), "__metadata__")

    meta <- entries[["__metadata__"]]
    expect_gt(meta[["offset"]], 0)
    expect_lt(meta[["offset"]], length(b))
    expect_gt(meta[["size"]],   0)
    expect_lte(meta[["size"]],  length(b) - meta[["offset"]])
  })
})


describe("ZIP compatibility — readable by standard ZIP tools", {

  it("stdlib reader lists all entries", {
    ctx <- local_archive()
    nm <- utils::unzip(ctx$archive, list = TRUE)$Name
    expect_true(all(c("__cozip__", "__metadata__", "a.txt", "b.bin") %in% nm))
  })

  it("all entries use STORE", {
    ctx <- local_archive()
    cd <- parse_cd(ctx$archive_bytes)
    for (e in cd) {
      expect_equal(e$compress_method, 0L,
                   info = sprintf("entry '%s' is not STORE", e$filename))
    }
  })

  it("no encryption, no data descriptor", {
    ctx <- local_archive()
    cd <- parse_cd(ctx$archive_bytes)
    for (e in cd) {
      expect_equal(bitwAnd(e$flag_bits, 0x01), 0L,
                   info = sprintf("'%s' encrypted", e$filename))
      expect_equal(bitwAnd(e$flag_bits, 0x08), 0L,
                   info = sprintf("'%s' has DD", e$filename))
    }
  })

  it("user payloads byte-exact", {
    ctx <- local_archive()
    out_dir <- file.path(ctx$tmp_path, "extract")
    dir.create(out_dir)
    utils::unzip(ctx$archive, files = c("a.txt", "b.bin"), exdir = out_dir)
    rec_a <- readBin(file.path(out_dir, "a.txt"), raw(), n = length(SMALL_CONTENT))
    rec_b <- readBin(file.path(out_dir, "b.bin"), raw(), n = length(MEDIUM_CONTENT))
    expect_identical(rec_a, SMALL_CONTENT)
    expect_identical(rec_b, MEDIUM_CONTENT)
  })
})


describe("stage_metadata — pure planner", {

  it("drops path column", {
    ctx <- local_archive()
    meta <- cozip::stage_metadata(ctx$input_table)$metadata
    expect_false("path" %in% names(meta))
  })

  it("adds offset and size as uint64", {
    ctx <- local_archive()
    meta <- cozip::stage_metadata(ctx$input_table)$metadata
    expect_identical(meta$schema$GetFieldByName("offset")$type$ToString(), "uint64")
    expect_identical(meta$schema$GetFieldByName("size")$type$ToString(),   "uint64")
  })

  it("preserves user extras", {
    ctx <- local_archive()
    meta <- cozip::stage_metadata(ctx$input_table)$metadata
    expect_identical(as.character(meta[["category"]]), c("text", "binary"))
  })

  it("canonical column order", {
    ctx <- local_archive()
    meta <- cozip::stage_metadata(ctx$input_table)$metadata
    expect_identical(names(meta)[1:3], c("name", "offset", "size"))
  })

  it("sizes match source lengths", {
    ctx <- local_archive()
    meta <- cozip::stage_metadata(ctx$input_table)$metadata
    sizes <- as.numeric(meta[["size"]])
    expect_equal(sizes, c(length(SMALL_CONTENT), length(MEDIUM_CONTENT)))
  })

  it("offsets strictly increasing", {
    ctx <- local_archive()
    meta <- cozip::stage_metadata(ctx$input_table)$metadata
    offsets <- as.numeric(meta[["offset"]])
    expect_true(all(diff(offsets) > 0))
  })

  it("returns paths aligned with metadata", {
    ctx <- local_archive()
    res <- cozip::stage_metadata(ctx$input_table)
    expect_equal(res$paths$name, c("a.txt", "b.bin"))
    expect_equal(res$paths$path, c(ctx$fixtures$small, ctx$fixtures$medium))
    expect_equal(nrow(res$paths), res$metadata$num_rows)
  })

  it("paths drive stage_create end-to-end", {
    ctx <- local_archive()
    res <- cozip::stage_metadata(ctx$input_table)
    pq  <- file.path(ctx$tmp_path, "meta.parquet")
    arrow::write_parquet(res$metadata, pq)
    out <- file.path(ctx$tmp_path, "out.zip")
    cozip::stage_create(out, res$paths, pq)
    assert_valid_cozip(readBin(out, raw(), n = file.size(out)))
  })

  it("rejects offset in input", {
    ctx <- local_archive()
    bad <- arrow::arrow_table(
      name   = "a.txt",
      path   = ctx$fixtures$small,
      offset = 0
    )
    expect_error(cozip::stage_metadata(bad), "reserved")
  })

  it("rejects duplicate names", {
    ctx <- local_archive()
    bad <- arrow::arrow_table(
      name = c("x", "x"),
      path = c(ctx$fixtures$small, ctx$fixtures$medium)
    )
    expect_error(cozip::stage_metadata(bad), "duplicate")
  })

  it("rejects reserved archive name", {
    ctx <- local_archive()
    bad <- arrow::arrow_table(
      name = "__metadata__",
      path = ctx$fixtures$small
    )
    expect_error(cozip::stage_metadata(bad), "reserved")
  })
})


describe("stage_create — embed user-written parquet verbatim", {

  write_meta <- function(ctx) {
    pq <- file.path(ctx$tmp_path, "meta.parquet")
    arrow::write_parquet(cozip::stage_metadata(ctx$input_table)$metadata, pq)
    pq
  }

  it("packs a valid archive", {
    ctx <- local_archive()
    pq  <- write_meta(ctx)
    out <- file.path(ctx$tmp_path, "out.zip")
    cozip::stage_create(out, make_paths_arg(ctx$input_table), pq)
    assert_valid_cozip(readBin(out, raw(), n = file.size(out)))
  })

  it("metadata parquet embedded verbatim", {
    ctx <- local_archive()
    pq  <- write_meta(ctx)
    original <- readBin(pq, raw(), n = file.size(pq))

    out <- file.path(ctx$tmp_path, "out.zip")
    cozip::stage_create(out, make_paths_arg(ctx$input_table), pq)

    arc      <- readBin(out, raw(), n = file.size(out))
    embedded <- extract_metadata_payload(arc)
    expect_identical(embedded, original)
  })

  it("rejects parquet with path column", {
    ctx <- local_archive()
    md_df <- as.data.frame(cozip::stage_metadata(ctx$input_table)$metadata)
    md_df$path <- c("/tmp/a", "/tmp/b")
    pq <- file.path(ctx$tmp_path, "meta.parquet")
    arrow::write_parquet(arrow::arrow_table(md_df), pq)

    expect_error(
      cozip::stage_create(file.path(ctx$tmp_path, "out.zip"),
                          make_paths_arg(ctx$input_table), pq),
      "path"
    )
  })

  it("rejects parquet missing required columns", {
    ctx <- local_archive()
    bad <- arrow::arrow_table(name = c("a.txt", "b.bin"))
    pq  <- file.path(ctx$tmp_path, "meta.parquet")
    arrow::write_parquet(bad, pq)

    expect_error(
      cozip::stage_create(file.path(ctx$tmp_path, "out.zip"),
                          make_paths_arg(ctx$input_table), pq),
      "required column"
    )
  })

  it("rejects offset mismatch", {
    ctx <- local_archive()
    bad <- arrow::arrow_table(
      name   = c("a.txt", "b.bin"),
      offset = arrow::Array$create(c(0, 0), type = arrow::uint64()),
      size   = arrow::Array$create(
        c(length(SMALL_CONTENT), length(MEDIUM_CONTENT)),
        type = arrow::uint64()
      )
    )
    pq <- file.path(ctx$tmp_path, "meta.parquet")
    arrow::write_parquet(bad, pq)

    expect_error(
      cozip::stage_create(file.path(ctx$tmp_path, "out.zip"),
                          make_paths_arg(ctx$input_table), pq),
      "offset mismatch"
    )
  })

  it("rejects row count mismatch", {
    ctx <- local_archive()
    bad <- arrow::arrow_table(
      name   = "a.txt",
      offset = arrow::Array$create(0, type = arrow::uint64()),
      size   = arrow::Array$create(0, type = arrow::uint64())
    )
    pq <- file.path(ctx$tmp_path, "meta.parquet")
    arrow::write_parquet(bad, pq)

    expect_error(
      cozip::stage_create(file.path(ctx$tmp_path, "out.zip"),
                          make_paths_arg(ctx$input_table), pq),
      "rows"
    )
  })
})


describe("schema metadata round-trips bit-perfect", {

  it("Arrow schema KV is preserved end-to-end", {
    ctx <- local_archive()

    geo_value <- paste0(
      '{"version":"1.0.0","primary_column":"geometry",',
      '"columns":{"geometry":{"encoding":"WKB","crs":null}}}'
    )

    tagged <- cozip::stage_metadata(ctx$input_table)$metadata$ReplaceSchemaMetadata(
      list(geo = geo_value, `asterisk:tag` = "cozip-conformance")
    )

    pq <- file.path(ctx$tmp_path, "meta.parquet")
    arrow::write_parquet(tagged, pq)

    out <- file.path(ctx$tmp_path, "out.zip")
    cozip::stage_create(out, make_paths_arg(ctx$input_table), pq)

    arc      <- readBin(out, raw(), n = file.size(out))
    embedded <- extract_metadata_payload(arc)

    rec_pq <- file.path(ctx$tmp_path, "recovered.parquet")
    writeBin(embedded, rec_pq)
    recovered <- arrow::read_parquet(rec_pq, as_data_frame = FALSE)

    md <- recovered$schema$metadata
    expect_identical(md[["geo"]],           geo_value)
    expect_identical(md[["asterisk:tag"]], "cozip-conformance")
  })
})


describe("GeoParquet round-trip via sf + geoarrow", {

  it("sf reads embedded GeoParquet with CRS, names, and coords intact", {
    skip_if_not_installed("sf")
    skip_if_not_installed("geoarrow")

    suppressPackageStartupMessages(library(geoarrow))
    on.exit(try(detach("package:geoarrow", unload = TRUE), silent = TRUE),
            add = TRUE)

    ctx <- local_archive()

    md_df <- as.data.frame(cozip::stage_metadata(ctx$input_table)$metadata)
    geom <- sf::st_sfc(
      sf::st_point(c(-77.04, -12.05)),
      sf::st_point(c( 2.35,  48.86)),
      crs = 4326
    )
    md_sf <- sf::st_sf(md_df, geometry = geom)

    pq <- file.path(ctx$tmp_path, "meta.parquet")
    arrow::write_parquet(md_sf, pq)

    out <- file.path(ctx$tmp_path, "out.zip")
    cozip::stage_create(out, make_paths_arg(ctx$input_table), pq)

    arc      <- readBin(out, raw(), n = file.size(out))
    embedded <- extract_metadata_payload(arc)

    rec_pq <- file.path(ctx$tmp_path, "recovered.parquet")
    writeBin(embedded, rec_pq)
    recovered <- arrow::read_parquet(rec_pq)
    recovered$geometry <- sf::st_as_sfc(recovered$geometry)

    expect_s3_class(recovered, "sf")
    expect_equal(sf::st_crs(recovered)$epsg, 4326L)
    expect_identical(recovered$name, c("a.txt", "b.bin"))

    coords <- sf::st_coordinates(recovered)
    expect_equal(unname(coords[1, "X"]), -77.04)
    expect_equal(unname(coords[1, "Y"]), -12.05)
    expect_equal(unname(coords[2, "X"]),   2.35)
    expect_equal(unname(coords[2, "Y"]),  48.86)
  })
})